File: neatformcore/src/main/java/com/opensrp/neatformcore/views/controls/CheckboxNFormView.java
package com.opensrp.neatformcore.views.controls;

import android.content.Context;
import android.support.v7.widget.AppCompatCheckBox;
import android.util.AttributeSet;

import com.opensrp.neatformcore.domain.model.NFormViewData;
import com.opensrp.neatformcore.domain.model.NFormViewOption;
import com.opensrp.neatformcore.domain.model.NFormViewProperty;
import com.opensrp.neatformcore.domain.view.DataPassListener;
import com.opensrp.neatformcore.domain.view.NFormView;
import com.opensrp.neatformcore.views.data.ViewDataHandler;

public class CheckboxNFormView extends AppCompatCheckBox implements NFormView {

    private static String TAG = CheckboxNFormView.class.getCanonicalName();

    private DataPassListener dataPassListener;

    private NFormViewProperty viewProperty;

    private NFormViewOption viewOption;

    public CheckboxNFormView(Context context) {
        super(context);
    }

    public CheckboxNFormView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CheckboxNFormView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public NFormViewOption getViewOption() {
        return this.viewOption;
    }

    @Override
    public NFormView initView(NFormViewProperty viewProperty, ViewDataHandler viewDataHandler) {
        this.viewOption = new NFormViewOption(this);
        setOnDataPassListener(viewDataHandler);
        return this;
    }

    @Override
    public NFormViewData getViewData() {
        return new NFormViewData();
    }

    @Override
    public void setOnDataPassListener(DataPassListener dataPassListener) {
        if (this.dataPassListener == null) {
            this.dataPassListener = dataPassListener;
        }
    }
}

File: neatformcore/src/main/java/com/nerdstone/neatformcore/views/widgets/CheckBoxNFormView.kt
package com.nerdstone.neatformcore.views.widgets

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.CheckBox
import com.nerdstone.neatformcore.domain.builders.ViewBuilder
import com.nerdstone.neatformcore.domain.data.DataActionListener
import com.nerdstone.neatformcore.domain.model.NFormViewData
import com.nerdstone.neatformcore.domain.model.NFormViewDetails
import com.nerdstone.neatformcore.domain.model.NFormViewProperty
import com.nerdstone.neatformcore.domain.view.NFormView
import com.nerdstone.neatformcore.domain.view.RootView
import com.nerdstone.neatformcore.domain.view.RulesHandler
import com.nerdstone.neatformcore.utils.ViewUtils
import com.nerdstone.neatformcore.views.builders.CheckBoxViewBuilder
import com.nerdstone.neatformcore.views.handlers.ViewDispatcher

class CheckBoxNFormView : CheckBox, NFormView {

    override var dataActionListener: DataActionListener? = null
    override val viewBuilder: ViewBuilder = CheckBoxViewBuilder(this)
    override val viewDetails: NFormViewDetails = NFormViewDetails(this)
    override lateinit var viewProperties: NFormViewProperty

    override val viewData: NFormViewData
        get() = NFormViewData()

    override val nFormRootView: RootView
        get() = this.parent as RootView

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    override fun initView(
        viewProperty: NFormViewProperty,
        viewDispatcher: ViewDispatcher
    ): NFormView = ViewUtils.setupView(this, viewProperty, viewBuilder, viewDispatcher)

    override fun setOnDataPassListener(dataActionListener: DataActionListener) {
        if (this.dataActionListener == null) this.dataActionListener = dataActionListener
    }


    override fun mapViewIdToName(rulesHandler: RulesHandler) {
        id = View.generateViewId()
        rulesHandler.viewIdsMap[this.viewProperties.name] = id
    }

    override fun setChecked(checked: Boolean) {
        super.setChecked(checked)
        dataActionListener?.also {
            if (checked) {
                this.viewDetails.value = mutableMapOf(viewDetails.name to text.toString())
            } else {
                this.viewDetails.value = mutableMapOf(viewDetails.name to null)
            }
            it.onPassData(viewDetails)
        }

    }
}